docs/design/vcpu-threads-pinning.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc for Kata Containers' VCPUs Pinning Feature
+
+## Background
+By now, vCPU threads of Kata Containers are scheduled randomly to CPUs. And each pod would request a specific set of CPUs which we call it CPU set (just the CPU set meaning in Linux cgroups).    
+
+If the number of vCPU threads are equal to that of CPUs claimed in CPU set, we can then pin each vCPU thread to one specified CPU, to reduce the cost of random scheduling. 
+
+## Detailed Design
+
+### Passing Config Parameters
+Two ways are provided to use this vCPU thread pinning feature: through qemu configuration file and through annotations. Finally the pinning parameter is passed to the HypervisorConfig struct.
+
+### Related Linux Thread Scheduling API
+
+| API Info          | Value                                                     |
+|-------------------|-----------------------------------------------------------|
+| Package           | golang.org/x/sys/unix                                     |
+| Method            | unix.SchedSetaffinity(thread_id, &unixCPUSet)             |
+| Official Doc Page | https://pkg.go.dev/golang.org/x/sys/unix#SchedSetaffinity |
+
+### When is VCPUs Pinning Checked?
+
+As shown in Section 1, when num(vCPU threads) == num(CPUs in CPU set), we shall pin each vCPU thread to a specified CPU. And when this condition is broken, we should restore to the original random scheduling pattern.  
+So when may num(CPUs in CPU set) change? There are 5 possible scenes:
+
+| Possible scenes                   | Related Code                               |
+|-----------------------------------|--------------------------------------------|
+| when creating a container         | File Sandox.go, in method CreateContainer  |
+| when starting a container         | File Sandox.go, in method StartContainer   |
+| when deleting a container         | File Sandox.go, in method DeleteContainer  |
+| when updating a container         | File Sandox.go, in method UpdateContainer  |
+| when creating multiple containers | File Sandox.go, in method createContainers |
+
+### Core Pinning Logics
```

Comment:
Please describe the core logic here, explicitly I'm missing:
1. How are the default vCPUs handled?
2. cpuset only comes from a container spec. How do we handle multiple containers with different cpusets?
3. How does this work from Kubernetes e2e point of view?

Reply:
Hi,

Reply:
Hi,

Reply:
Hi @bergwolf 
1. I didn't take default vCPU logics into consideration in this case. I only checks if the number of vCPU threads and the number of CPU requested are equal, and I need the number of default vCPU threads in this case, by using method "s.hypervisor.GetThreadIDs(ctx)". I think this method has taken default vCPU logics into consideration.
2. CPUSets are acquired by using method "s.getSandboxCPUSet()". In this method the sandbox goes through CPUSet of each container and returns a final overall CPUSet.
3. Two ways are provided to use this feature. By default the switch of this feature is passed by line "enable_vcpus_pinning = ture" in file "configuration-qemu.toml.in". Also this switch can be passed by annotation configuration, in file "oci/utils.go" "vcAnnotations.EnableVCPUsPinning".

Reply:
Hi if you have any other questions just let me know and I'll fix them and make some changes to the code.